### PR TITLE
Feature/oracle support

### PR DESCRIPTION
**Pull request by Curlack**
This covers all but one task. Not sure what to do for "Create a derived `OracleManagedTestProvider` class to derive from...".
These changes are for baseline Oracle setup. To confirm setup and build scripts work I ran `PetaPoco.Tests.Integration.DatabaseTests.Construct_GivenConnection_ShouldBeValid`, which executed successfully.

### CODE REVIEW

**Comment by Ste1io** [comment]
Looks fantastic based on a quick fly-by over the commit @Curlack. I'll take a look at everything between now and tomorrow.

File: PetaPoco/Utilities/ParametersHelper.cs [hunk 1]
```diff
@@ -170,9 +170,8 @@ void ProcessArg(object arg)
                 else
                 {
                     var type = arg.GetType();
-                    // TODO: Include second param `paramName: nameof(args)` in thrown ArgumentException
                     if (type.IsValueType || type == typeof(string))
-                        throw new ArgumentException($"Value type or string passed as stored procedure argument: {arg}");
+                        throw new ArgumentException($"Value type or string passed as stored procedure argument: {arg}", nameof(args));
```
Comment (Ste1io) [inline PetaPoco/Utilities/ParametersHelper.cs:174]: Cheers, thanks for this. Added this comment while rewriting the xml docs and was trying not to let myself get distracted lol.

**Review by Ste1io — COMMENTED** [review]
I'm curious what the script parsing bug was? Any chance it was related to https://github.com/CollaboratingPlatypus/PetaPoco/commit/7e6c9401adb011fd8f4b5f07e8210001da529f2b?

**Comment by Curlack** [comment]
When I reused that part of the code, to also cater for comments in the
script, I didn't realize that the semi colon was required.
During troubleshooting, I found that the next line was jammed to the back
of the comment seperated by '\r\n'. Then I realized the code should rather
check lines within semi-colon seperated block of text. Since that was gonna
fix my issue, I also improved the part of the code I borrowed it from.
Don't think it is related though.

File: PetaPoco.Tests.Integration/Providers/OracleTestProvider.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Linq;
+
+/* Converted SqlServerBuildDatabase.sql scripts using MSSQLTips.com to modify datatypes like for like
+ * Link: https://www.mssqltips.com/sqlservertip/2944/comparing-sql-server-and-oracle-datatypes/#:~:text=SQL%20Server's%20is%20called%20XML,can%20handle%20up%20to%206GB.
+ * 
+ */
```
Comment (Ste1io) [inline PetaPoco.Tests.Integration/Providers/OracleTestProvider.cs:7]: Check typo in link and preceding comment line.

```suggestion
/* 
 * Converted build scripts from SqlServerBuildDatabase.sql using MSSQLTips.com for datatype conversions
 * Link: https://www.mssqltips.com/sqlservertip/2944/comparing-sql-server-and-oracle-datatypes
 */
```

**Comment by Ste1io** [comment]
Everything looks good @Curlack. Just one non-code-related typo, which isn't me attempting to be nitpicky I promise. About the `OracleManagedTestProvider`, that really isn't necessary until/if another provider is added. That was mainly in light of the DbProviderFactory for Oracle.

**Comment by Curlack** [comment]
Excellent, thanks. Committed your suggestion. Not used to aal these github
features lol. Will learn I suppose.

**Comment by Ste1io** [comment]
> Not used to aal these github features lol. Will learn I suppose.

I can relate...always afraid it will make some assumption I don't want and I'll regret it later (like the "Pull" defaulting to Fetch/Merge without so much as a courtesy message), so tend to using VSCode integration or CLI directly in most cases.

**Comment by Ste1io** [comment]
> NOTE: Added SuppressTfmSupportBuildWarnings in project file to circumvent "System.Runtime.CompilerServices.Unsafe doesn't support netcoreapp2.1" build error. Not sure how else to fix this issue.

I checked out your branch and looking into this...with the warning suppressed, it builds but the test fails when ran under dotnet2.1 target. The offending package causing that error message is `Oracle.ManagedDataAccess`; which is odd that it doesn't affect net472, seeing as that dependency is considered the most current iirc.

For reference:

```
2>------ Rebuild All started: Project: PetaPoco.Tests.Integration, Configuration: Debug Any CPU ------
2>C:\Users\there\.nuget\packages\system.runtime.compilerservices.unsafe\6.0.0\buildTransitive\netcoreapp2.0\System.Runtime.CompilerServices.Unsafe.targets(4,5): error : System.Runtime.CompilerServices.Unsafe doesn't support netcoreapp2.1. Consider updating your TargetFramework to netcoreapp3.1 or later.
2>Done building project "PetaPoco.Tests.Integration.csproj" -- FAILED.
```

**Comment by Curlack** [comment]
FYI, I wasn't able to build even before adding the Oracle nuget package.
Would like to know or try to figure out what to do so my build doesn't fail
without that change. For the time being used it like that to concentrate on
the important parts (I get distracted very easy lol). Thought of not
including it in the PR, but figured you'd guide in the right direction if I
do include it, 'no harm no foul' kind of thing.

**Comment by Ste1io** [comment]
> FYI, I wasn't able to build even before adding the Oracle nuget package.
Would like to know or try to figure out what to do so my build doesn't fail
without that change.

Can you copy the output log in here for me with the suppression and Oracle.ManagedDataAccess package removed from the project file?

> For the time being used it like that to concentrate on
the important parts

Completely understand, I'm the same. Hence the TODO you found ;)

> but figured you'd guide in the right direction if I
do include it, 'no harm no foul' kind of thing.

No harm done.

**Comment by Ste1io** [comment]
> The offending package causing that error message is Oracle.ManagedDataAccess; which is odd that it doesn't affect net472, seeing as that dependency is considered the most current iirc.

OK, I was wrong here; [Oracle.ManagedDataAccess](https://www.nuget.org/packages/Oracle.ManagedDataAccess/21.12.0?_src=template#supportedframeworks-body-tab) is the framework version (hence no issues with `net472` target). [Oracle.ManagedDataAccess.Core](https://www.nuget.org/packages/Oracle.ManagedDataAccess.Core/3.21.120?_src=template#supportedframeworks-body-tab) is the "modern" package for core apps (click links for compatible versions on nuget).

I'm adding the other and adding target conditions to both; this may require a change to the provider resolution method in the main library, will know shortly.

**Comment by Ste1io** [comment]
Screw it, tests are getting updated to `netcoreapp3.1`.  I'm running the full test suite rn, and will push the targetframeworks change in a separate PR to this branch to isolate it from these changes.

1697789162-PetaPoco_-_Microsoft_Visual_Studio

File: PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj [hunk 3]
```diff
@@ -70,6 +75,7 @@
     <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="2.1.0" Condition="'$(TargetFramework)' == 'netcoreapp2.1'" />
     <PackageReference Include="Moq" Version="[4.8.2,4.20)" />
     <PackageReference Include="Newtonsoft.Json" Version="13.0.3" />
+    <PackageReference Include="Oracle.ManagedDataAccess" Version="21.12.0" />
```
Comment (Ste1io) [inline PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj:78]:
```suggestion
    <PackageReference Include="Oracle.ManagedDataAccess" Version="21.12.0" Condition="'$(TargetFramework)' != 'netcoreapp2.1'" />
```

File: PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj [hunk 4]
```diff
@@ -11,6 +11,7 @@
 
     <IsPackable>false</IsPackable>
     <IsTestProject>true</IsTestProject>
+    <SuppressTfmSupportBuildWarnings>true</SuppressTfmSupportBuildWarnings>
```
Comment (Ste1io) [inline PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj:14]:
```suggestion
```

**Comment by Curlack** [comment]
Sorry dude, not in front of my laptop atm. Had a good laugh after reading
you comments.

Does that mean you don't need the output logs from me anymore?

Cause I'll only be able to give them to you this evening when I get home.

**Review by Ste1io — CHANGES_REQUESTED** [review]
Let's remove the suppression, and add a condition for the package. Oracle tests won't pass for netcoreapp2.1, but it should build fine (ie, no change from current state). Sets it up so we can roll in the 2.1->3.1 upgrade smoothly though.

**Review by Ste1io — APPROVED** [review]

**Comment by Ste1io** [comment]
> Had a good laugh after reading you comments.

Inline code comments or my half hazard PR thread comments? haha

> Does that mean you don't need the output logs from me anymore?

No, I definitely still want to see them if possible. Call in sick and get to your pc stat.
...Joking lol. Whenever you're able is fine, I'm in no rush. Just curious what was causing your build issue you mentioned so that's not an issue moving forward (for yourself, or possibly others).

**Review by Ste1io — CHANGES_REQUESTED** [review]
When you're at your pc and sync'd up with the changes here locally, @Curlack, could you rebase your feature branch onto `upstream/feature/oracle-support`, then add `Oracle.ManagedDataAccess.Core` package conditional to netcoreapp3.1 to the integration tests project file? Once that's committed/pushed this pr will be all set.

**Comment by Curlack** [comment]
> Can you copy the output log in here for me with the suppression and Oracle.ManagedDataAccess package removed from the project file?

Not sure if I was imagining things, but with or without the library, build is successful. Unfortunately I have no logs for you...

File: PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj [hunk 5]
```diff
@@ -53,6 +53,9 @@
     <Compile Remove="Databases\MSAccessTests\**" />
     <Compile Remove="Providers\MSAccessTestProvider.cs" />
     <EmbeddedResource Remove="Scripts\MSAccessBuildDatabase.sql" />
+    <PackageReference Include="Oracle.ManagedDataAccess.Core">
+      <Version>3.21.120</Version>
+    </PackageReference>
```
Comment (Ste1io) [inline PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj:58]: Given the current structure of this project file, this probably should be dropped in with the other package references IMO. It seems there's no right way of laying out these files, but grouping this with the list removal items is a sure-fire way of confusing yours truly some late night eventually. 😆

**Review by Ste1io — APPROVED** [review]

**Comment by Curlack** [comment]
When I looked at the file I thought the same thing, but I didn't put it
there. After installing the nuget package, it was like that all by itself
lol.

**Comment by Curlack** [comment]
Thanks

**Comment by Ste1io** [comment]
> When I looked at the file I thought the same thing, but I didn't put it there. After installing the nuget package, it was like that all by itself lol.

No worries. I'm set to merge soon as that's sorted. Any issues building at this point?

**Comment by Curlack** [comment]
No issues. Smooth sailing all the way. I had to install the 3.1 sdk and
runtime though.

File: PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj (hunk 5 above)
Comment (Curlack) [reply PetaPoco.Tests.Integration/PetaPoco.Tests.Integration.csproj]: All set.

**Comment by Ste1io** [comment]
Looks fantastic @Curlack. Let's get this Oracle merged in and see how he fairs with the tests :)

Merging Oracle 03

**Pull request merged**